### PR TITLE
Fix some typing in _TraceDict

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Chipping away towards #4354

### CODE REVIEW

File: pymc3/distributions/posterior_predictive.py [hunk 1]
```diff
@@ -50,21 +52,14 @@
 #    test_mixture_random_shape::test_mixture_random_shape
 #
 
-PosteriorPredictiveTrace = Dict[str, np.ndarray]
```
Comment (MarcoGorelli) [inline pymc3/distributions/posterior_predictive.py:53]: unused

File: pymc3/distributions/posterior_predictive.py [hunk 2]
```diff
@@ -50,21 +52,14 @@
...
 
 
 class HasName(Protocol):
     name: str
 
 
-if TYPE_CHECKING:
-    _TraceDictParent = UserDict[str, np.ndarray]
-else:
-    _TraceDictParent = UserDict
```
Comment (MarcoGorelli) [inline pymc3/distributions/posterior_predictive.py:64]: this strikes me as unnecessarily complicated, we can just annotate `data` directly in `_TraceDict`

File: pymc3/distributions/posterior_predictive.py [hunk 3]
```diff
@@ -75,24 +70,24 @@ class _TraceDict(_TraceDictParent):
...
-        point_list: Optional[List[Dict[str, np.ndarray]]] = None,
+        point_list: Optional[List[Point]] = None,
         multi_trace: Optional[MultiTrace] = None,
-        dict: Optional[Dict[str, np.ndarray]] = None,
+        dict_: Optional[Point] = None,
     ):
         """"""
         if multi_trace:
-            assert point_list is None and dict is None
-            self.data = {}  # Dict[str, np.ndarray]
```
Comment (MarcoGorelli) [inline pymc3/distributions/posterior_predictive.py:89]: this is already annotated above (`data: Point`), no need to re-annotate it here. `reveal_type` shows it's picked up correctly anyway

File: pymc3/distributions/posterior_predictive.py [hunk 4]
```diff
@@ -75,24 +70,24 @@ class _TraceDict(_TraceDictParent):
...
-    data: Dict[str, np.ndarray]
+    data: Point
 
     def __init__(
         self,
-        point_list: Optional[List[Dict[str, np.ndarray]]] = None,
+        point_list: Optional[List[Point]] = None,
         multi_trace: Optional[MultiTrace] = None,
-        dict: Optional[Dict[str, np.ndarray]] = None,
+        dict_: Optional[Point] = None,
```
Comment (MarcoGorelli) [inline pymc3/distributions/posterior_predictive.py:79]: renaming this as `dict` is a built-in

File: pymc3/distributions/posterior_predictive.py [hunk 5]
```diff
@@ -114,18 +109,18 @@ def arr_for(val):
...
+            self.varnames = list(dict_.keys())
+            self._len = dict_[self.varnames[0]].shape[0]
         assert self.varnames is not None and self._len is not None and self.data is not None
 
     def __len__(self) -> int:
         return self._len
 
-    def _extract_slice(self, slc: slice) -> "_TraceDict":
-        sliced_dict: Dict[str, np.ndarray] = {}
+    def _extract_slice(self, slc: slice) -> _TraceDict:
```
Comment (MarcoGorelli) [inline pymc3/distributions/posterior_predictive.py:122]: if we do `from __future__ import annotations` at the top then this works

File: pymc3/distributions/posterior_predictive.py [hunk 6]
```diff
@@ -302,12 +297,10 @@ def extend_trace(self, trace: Dict[str, np.ndarray]) -> None:
...
-        if keep_size:
-            return {
-                k: ary.reshape((nchains, ndraws, *ary.shape[1:])) for k, ary in ppc_trace.items()
-            }
-        # this gets us a Dict[str, np.ndarray] instead of my wrapped equiv.
-        return ppc_trace.data
+    if keep_size:
+        return {k: ary.reshape((nchains, ndraws, *ary.shape[1:])) for k, ary in ppc_trace.items()}
+    # this gets us a Dict[str, np.ndarray] instead of my wrapped equiv.
+    return ppc_trace.data
```
Comment (MarcoGorelli) [inline pymc3/distributions/posterior_predictive.py:303]: moving this outside from the context manager so mypy recognises that the return statement is always reached

**Review by michaelosthege — APPROVED** [review]
Yes, that looks very good!

Keep in mind that the days of the posterior predictive sampling code are numbered ([Timeline](https://github.com/pymc-devs/pymc3/wiki/Timeline)). With the transition to `RandomVariable` most of this code will go away. So type hinting contributions in Theano-PyMC (https://github.com/pymc-devs/Theano-PyMC/issues/200) could potentially improve the overall mypy experience much faster.

**Comment by MarcoGorelli** [comment]
Thanks - it's not really clear to me which of the pymc3 internals will be replaced by `RandomVariable` though

**Pull request merged**